File: src/main/java/xyz/oribuin/lilori/managers/music/TrackScheduler.java
package xyz.oribuin.lilori.managers.music;

import com.sedmelluq.discord.lavaplayer.player.AudioPlayer;
import com.sedmelluq.discord.lavaplayer.player.event.AudioEventAdapter;
import com.sedmelluq.discord.lavaplayer.track.AudioTrack;
import com.sedmelluq.discord.lavaplayer.track.AudioTrackEndReason;
import xyz.oribuin.lilori.commands.music.CmdLoop;

import java.util.LinkedList;
import java.util.Queue;

public class TrackScheduler extends AudioEventAdapter {
    private final AudioPlayer player;
    private final Queue<AudioTrack> queue;

    public TrackScheduler(AudioPlayer player) {
        this.player = player;
        this.player.setVolume(75);
        this.queue = new LinkedList<>();
    }

    public void queue(AudioTrack track) {
        if (!player.startTrack(track, true)) {
            queue.offer(track);
        }
    }

    public void nextTrack() {
        player.startTrack(queue.poll(), true);
    }

    @Override
    public void onTrackEnd(AudioPlayer player, AudioTrack audioTrack, AudioTrackEndReason endReason) {

        if (endReason.mayStartNext)
            if (new CmdLoop().isLooping()) {
                player.startTrack(audioTrack, false);
            } else {
                this.nextTrack();
            }

    }


}

File: src/main/kotlin/xyz/oribuin/lilori/managers/music/TrackScheduler.kt
package xyz.oribuin.lilori.managers.music

import com.sedmelluq.discord.lavaplayer.player.AudioPlayer
import com.sedmelluq.discord.lavaplayer.player.event.AudioEventAdapter
import com.sedmelluq.discord.lavaplayer.track.AudioTrack
import com.sedmelluq.discord.lavaplayer.track.AudioTrackEndReason
import xyz.oribuin.lilori.LilOri
import java.util.*

class TrackScheduler(private val bot: LilOri, private val player: AudioPlayer) : AudioEventAdapter() {
    private val queue: Queue<AudioTrack>
    fun queue(track: AudioTrack, override: Boolean) {
        if (!player.startTrack(track, !override)) {
            queue.offer(track)
        }
    }

    fun nextTrack() {
        player.startTrack(queue.poll(), true)
    }

    override fun onTrackEnd(player: AudioPlayer, audioTrack: AudioTrack, endReason: AudioTrackEndReason) {
        if (endReason.mayStartNext)
            nextTrack()
    }

    init {
        player.volume = 75
        queue = LinkedList()
    }
}